src/core/unit.c

```diff
@@ -727,6 +728,9 @@ void unit_free(Unit *u) {
 
         bpf_program_unref(u->bpf_device_control_installed);
 
+        (void) cgroup_bpf_detach_programs(u, u->allow_bind_progs);
```

Comment:
this doesn't look right.

unit_free() is called whenever we free a unit, but that happens during PID1 reexec too for example. and during reexec we shouldn't detach the prog i figure.

(we had some issues with the other bpf progs on this where we momentarily dropped them during reload/reexec. But the goal really needs to be to pass the fd over the reexec)

Reply:
I'm also getting a `Failed to detach program from cgroup '/sys/fs/cgroup': No such file or directory` error on this line. I thought it was because it is called after `unit_release_cgroup`, I tried to move it before but I still get the same message.

Reply:
@mauriciovasquezbernal

Cgroup may also be released in [unit_prune_cgroup](https://github.com/systemd/systemd/blob/main/src/core/service.c#L1100#L1101) which calling  `unit_release_cgroup`, which is why u->cgroup_path is `NULL` by the time `unit_free` happens.

Reply:
So according to https://github.com/systemd/systemd/pull/17495 we shouldn't detach the program here?

Reply:
@mauriciovasquezbernal

Programs should be detached in two cases:
* An option with an empty value, e.g. `AllowBindPorts=` is passed 
* A cgroup is destroyed

IMO the smaller the gap between programs detach and cgroup's `rmdir` the better. `unit_free` is responsible for cleaning up logical state (structures associated with a unit). 
Looking at [execute.c](https://github.com/systemd/systemd/blob/main/src/core/execute.c), that seems to be more appropriate place.

Reply:
@mauriciovasquezbernal @poettering 

Made bpf programs attachments persist through daemon re{load|exec}.
[bpf_link](https://lore.kernel.org/netdev/87k143t682.fsf@toke.dk/T/) abstraction worked like a charm: closing the fd associated with bpf_link is enough to detach linked bpf program from a cgroup. To survive daemon-re{load|exec} bpf_link fd should be serialized and closed after newly loaded program is attached.
That is more elegant than detaching with `bpf_prog_detach`, i.e. bpf_allow_multi flag is implied, no need to close program fd after detach and worry about attach types.

Downside is that won't work on older kernels.